### ISSUE TITLE
Incorrect dependent template name error

### ISSUE THREAD

**Issue by llvmbot** [llvm/llvm-project#17775]
|  |  |
| --- | --- |
| Bugzilla Link | [17401](https://llvm.org/bz17401) |
| Version | 3.3 |
| OS | All |
| Reporter | LLVM Bugzilla Contributor |
| CC | @DougGregor |

## Extended Description 
In the code shown below, Clang claims that the call to `foo` within `C::baz()` is template-dependent and therefore must be prefixed with the `template` keyword.  I'm not a standards expert, but it does not appear to me to be template-dependent -- the call clearly always resolves to `A::foo()`.  Moreover, if the type of `C::b` is changed to `B*` and the call becomes `b->bar().foo<T>()` then no error is reported, which seems pretty suspicious.

Tested on Clang 3.2 (Ubuntu) and Clang 3.3 (Xcode 5).

G++ (4.7, 4.8) does not report an error on this code.
```
$ cat test.c++
```
```cpp
class A {
public:
  template <typename T>
  T foo();
};

class B {
public:
  A bar();
};

template <typename T>
class C {
public:
  T baz() {
    return b.bar().foo<T>();
  }
  
private:
  B b;
};
```
```
$ clang++ -c test.c++
test.c++:16:20: error: use 'template' keyword to treat 'foo' as a dependent template name
    return b.bar().foo<T>();
                   ^
                   template 
1 error generated.
$
```

**Comment by llvmbot** [comment]
@llvm/issue-subscribers-clang-frontend

Author: None (llvmbot)

<details>
|  |  |
| --- | --- |
| Bugzilla Link | [17401](https://llvm.org/bz17401) |
| Version | 3.3 |
| OS | All |
| Reporter | LLVM Bugzilla Contributor |
| CC | @DougGregor |

## Extended Description 
In the code shown below, Clang claims that the call to `foo` within `C::baz()` is template-dependent and therefore must be prefixed with the `template` keyword.  I'm not a standards expert, but it does not appear to me to be template-dependent -- the call clearly always resolves to `A::foo()`.  Moreover, if the type of `C::b` is changed to `B*` and the call becomes `b-&gt;bar().foo&lt;T&gt;()` then no error is reported, which seems pretty suspicious.

Tested on Clang 3.2 (Ubuntu) and Clang 3.3 (Xcode 5).

G++ (4.7, 4.8) does not report an error on this code.
```
$ cat test.c++
```
```cpp
class A {
public:
  template &lt;typename T&gt;
  T foo();
};

class B {
public:
  A bar();
};

template &lt;typename T&gt;
class C {
public:
  T baz() {
    return b.bar().foo&lt;T&gt;();
  }
  
private:
  B b;
};
```
```
$ clang++ -c test.c++
test.c++:16:20: error: use 'template' keyword to treat 'foo' as a dependent template name
    return b.bar().foo&lt;T&gt;();
                   ^
                   template 
1 error generated.
$
```
</details>

**Comment by shafik** [comment]
Related bug: https://github.com/llvm/llvm-project/issues/37647

**Comment by shafik** [comment]
Looks fixed in clang10: https://godbolt.org/z/E91qMzf14
